components/net/filemanager_thread.rs

```diff
@@ -0,0 +1,145 @@
+    receiver: IpcReceiver<FileManagerThreadMsg>,
+    idmap: RefCell<HashMap<Uuid, String>>,
+}
+
+enum SelectedPort {
+    Single(IpcSender<FileManagerResult<(Uuid, PathBuf, u64)>>),
+    Multiple(IpcSender<FileManagerResult<Vec<(Uuid, PathBuf, u64)>>>)
+}
+
+impl FileManager {
+    fn new(recv: IpcReceiver<FileManagerThreadMsg>) -> FileManager {
+        FileManager {
+            receiver: recv,
+            idmap: RefCell::new(HashMap::new()),
+        }
+    }
+
+    fn new_thread() -> IpcSender<FileManagerThreadMsg> {
+        let (chan, recv) = ipc::channel().unwrap();
+
+        spawn_named("FileManager".to_owned(), move || {
+            FileManager::new(recv).start();
+        });
+
+        chan
+    }
+
+    /// Start the file manager event loop
+    pub fn start(&mut self) {
+        loop {
+            match self.receiver.recv().unwrap() {
+                FileManagerThreadMsg::SelectFile(sender) => self.select(SelectedPort::Single(sender)),
+                FileManagerThreadMsg::SelectFiles(sender) => self.select(SelectedPort::Multiple(sender)),
+                FileManagerThreadMsg::ReadFile(sender, id) => self.read_file(sender, id),
+                FileManagerThreadMsg::DeleteFileID(id) => self.delete_fileid(id),
+            }
+        }
+    }
+}
+
+impl FileManager {
+    fn select(&mut self, sender: SelectedPort) {
+        // TODO: Pull the dialog UI in and get selected
+        match sender {
+            SelectedPort::Multiple(sender) => {
+                let selected_paths = vec![""];
+
+                let mut replies = vec![];
+
+                for path in selected_paths {
+                    match self.create_entry(path) {
+                        Some(triple) => replies.push(triple),
+                        None => {
+                            let _ = sender.send(Err(FileManagerThreadError::InvalidSelection));
+                            return;
+                        }
+                    }
+                }
+
+                let _ = sender.send(Ok(replies));
+            },
+            SelectedPort::Single(sender) => {
+                let selected_path = "";
+
+                match self.create_entry(selected_path) {
+                    Some(triple) => {
+                        let _ = sender.send(Ok(triple));
+                    },
+                    None => {
+                        let _ = sender.send(Err(FileManagerThreadError::InvalidSelection));
+                    }
+                }
+            }
+        }
+    }
+
+    fn create_entry(&mut self, file_path: &str) -> Option<(Uuid, PathBuf, u64)> {
+        match File::open(file_path) {
+            Ok(handler) => {
+                let id = Uuid::new_v4();
+                self.idmap.borrow_mut().insert(id, file_path.to_string());
+
+                // Unix Epoch: https://doc.servo.org/std/time/constant.UNIX_EPOCH.html
+                let epoch = handler.metadata().and_then(|metadata| metadata.modified()).map_err(|_| ())
```

Comment:
probably should leave a note about mtime vs ctime, iirc we haven't made that decision yet